ta_services/secure_key_services/include/sks_ta.h

```diff
@@ -14,13 +14,615 @@
+#define SKS_PROC_GENERATE		BIT(8UL)
+#define SKS_PROC_GENERATE_PAIR		BIT(9UL)
+#define SKS_PROC_WRAP			BIT(10UL)
+#define SKS_PROC_UNWRAP			BIT(11UL)
+#define SKS_PROC_DERIVE			BIT(12UL)
+
+/*
+ * SKS_CMD_CK_INIT_TOKEN - Initialiaze PKCS#11 token
+ *
+ * param#0: in(*)-memref : [uint32_t slot_id]
+ *			   [uint32_t pin_len]
+ *			   [uint8_t pin[pin_len]]
+ *			   [uint8_t label[32]]
+ * param#1: none
+ * param#2: none
+ * param#3: none
+ *
+ * This commands relates to the PKCS#11 API function C_InitToken().
+ */
+#define SKS_CMD_CK_INIT_TOKEN		0x00000006
+
+struct sks_item_length {
+	uint32_t byte_size;
+};
+
+/*
+ * SKS_CMD_CK_INIT_PIN - Initialiaze PKCS#11 token PIN
+ *
+ * param#0: in(*)-memref : [uint32_t session_handle]
+ *			   [uint32_t pin_len]
+ *			   [uint8_t pin[pin_len]]
+ * param#1: none
+ * param#2: none
+ * param#3: none
+ *
+ * This commands relates to the PKCS#11 API function C_InitPIN().
+ */
+#define SKS_CMD_CK_INIT_PIN		0x00000007
+
+struct sks_handle {
+	uint32_t handle;
+};
+
+/*
+ * SKS_CMD_CK_SET_PIN - Set PKCS#11 token PIN
+ *
+ * param#0: in(*)-memref : [uint32_t session_id]
+ *			   [uint32_t old_pin_len]
+ *			   [uint8_t old_pin[old_pin_len]]
+ *			   [uint32_t new_pin_len]
+ *			   [uint8_t new_pin[new_pin_len]]
+ * param#1: none
+ * param#2: none
+ * param#3: none
+ *
+ * This commands relates to the PKCS#11 API function C_SetPIN()
+ */
+#define SKS_CMD_CK_SET_PIN		0x00000008
+
+/*
+ * SKS_CMD_CK_OPEN_RO_SESSION - Open Read-only Session
+ *
+ * param#0: in(*)-memref : [uint32_t slot_id]
+ * param#1: none
+ * param#2: out-memref : [uint32_t session_handle]
+ * param#3: none
+ *
+ * This commands relates to the PKCS#11 API function C_OpenSession() for a
+ * read-only session.
+ */
+#define SKS_CMD_CK_OPEN_RO_SESSION	0x00000009
+
+/*
+ * SKS_CMD_CK_OPEN_RW_SESSION - Open Read/Write Session
+ *
+ * param#0: in(*)-memref : [uint32_t slot_id]
+ * param#1: none
+ * param#2: out-memref : [uint32_t session_handle]
+ * param#3: none
+ *
+ * This commands relates to the PKCS#11 API function C_OpenSession() for a
+ * read/Write session.
+ */
+#define SKS_CMD_CK_OPEN_RW_SESSION	0x0000000a
+
+/*
+ * SKS_CMD_CK_CLOSE_SESSION - Open Read/Write Session
+ *
+ * param#0: in(*)-memref : [uint32_t session_handle]
+ * param#1: none
+ * param#2: none
+ * param#3: none
+ *
+ * This commands relates to the PKCS#11 API function C_CloseSession().
+ */
+#define SKS_CMD_CK_CLOSE_SESSION	0x0000000b
+
+/*
+ * SKS_CMD_CK_SESSION_INFO - Get Cryptoki information on a session
+ *
+ * param#0: in(*)-memref : [uint32_t session_handle]
+ * param#1: none
+ * param#2: out-memref : [struct sks_ck_session_info]
+ * param#3: none
+ *
+ * This commands relates to the PKCS#11 API function C_GetSessionInfo().
+ */
+#define SKS_CMD_CK_SESSION_INFO		0x0000000c
+
+struct sks_ck_session_info {
+	uint32_t slot_id;
+	uint32_t state;
+	uint32_t flags;
+	uint32_t error_code;
+};
+
+/*
+ * SKS_CMD_CK_CLOSE_ALL_SESSIONS - Close all slot's pending sessions
```

Comment:
What does "pending"session means ?

> From PKCS#11 standard:
> C_CloseAllSessions closes all sessions an application has with a token.

Reply:
You're right. I have fixed the description.
My early developments around this are buggy. I will fix that. Thanks.